Seojin validate fqdn

SeojLee opened the pull request:
<!-- kody-pr-summary:start -->
This pull request implements a significant change in how host identifiers are generated and validated within the system, moving from IP address-based IDs to FQDN-based CRC32 hashes.

Key changes include:

1.  **Transition to FQDN-based Host IDs**:
    *   Host IDs (`csv2_host_id`, `local_host_id`, and those derived from FQDNs in general) are now calculated using a CRC32 hash of the Fully Qualified Domain Name (FQDN) instead of an integer representation of an IPv4 address. This change is applied in `lib/db_config.py` for core host ID generation and in `lib/view_utils.py` for FQDN input processing.
    *   Several utility functions related to IP address-based host ID conversions (`get_host_id_by_ip_or_tag`, `get_host_ip_by_host_id`, `get_host_tag_by_host_id`) have been removed from `lib/db_config.py`.

2.  **Enhanced FQDN Update Validation**:
    *   A new validation mechanism has been introduced in `web_frontend/cloudscheduler/csv2/group_views.py` when updating a group's `htcondor_fqdn`.
    *   If a user attempts to change the `htcondor_fqdn` for a group, the system now calculates the new `htcondor_host_id` using the CRC32 hash of the submitted FQDN.
    *   Before allowing the update, it performs a critical check: it verifies if the *old* `htcondor_host_id` (derived from the previous FQDN) is still referenced in active `condor_jobs`, `condor_machines`, `condor_worker_gsi`, or `csv2_sevice_catalog` tables.
    *   If the old host ID is found in any of these active components, the FQDN update is rejected, and an error message is returned (e.g., "Group update failed - host_id found in active jobs"), preventing potential data inconsistencies or operational issues.

These changes aim to improve the robustness and integrity of host identification by using stable FQDN hashes and to prevent data corruption when FQDNs are updated for active groups.
<!-- kody-pr-summary:end -->

SeojLee committed:
added FQDN validation to Group Config

SeojLee committed:
added additional check for service catalog

colsond commented:
The functionality here will have to be reworked because the problem it's trying to solve is more widespread.

SeojLee committed:
pushed changes for rebuild

SeojLee committed:
Remove unused Crc32 import and checksum calculation

Removed unused Crc32 checksum calculation for submitted FQDN.

SeojLee committed:
Replace IP address assignment with CRC32 calculation

SeojLee committed:
Calculate htcondor_host_id from submitted FQDN

Add logic to calculate htcondor_host_id from submitted_fqdn

SeojLee committed:
Import Crc32 for CRC calculations

SeojLee committed:
Implement cleanup of stale service catalog entries

Added a function to clean up stale service catalog entries based on active IDs and a configured timeout.

SeojLee committed:
consistent casing

SeojLee committed:
Add where clause to db_query for stale entries

Commit:
fix indentation errors

colsond commented on cloudscheduler/csmain:
```diff
@@ -125,6 +125,47 @@ def available_resources_initialize(config):
 
     return used_resources
 
+def cleanup_stale_service_catalog(config):
```
restrict cleanup to only condor poller services because everything else should be running on the csv2 host which is very unlikely to change ID

colsond commented on cloudscheduler/csmain:
```diff
@@ -125,6 +125,47 @@ def available_resources_initialize(config):
+                logging.info(entry)
+                host_id = entry.get(col)
+                if host_id and host_id not in active_ids:
+                    config.db_execute("delete from %s where %s = %s" % (table, col, host_id))
```
would be best to batch all the deletes and do them all at the end of the loop.

colsond commented on lib/db_config.py:
```diff
@@ -75,12 +77,11 @@ def __init__(self, db_yaml, categories, db_config_dict=False, db_config_only=Fal
-        self.local_host_id = int(ipaddress.IPv4Address(socket.gethostbyname(socket.gethostname())))
+        self.local_hostname_fqdn = socket.getfqdn()
+        self.local_host_id = int(Crc32.calc(self.local_hostname_fqdn.encode("utf-8")))
+        self.csv2_host_id = int(Crc32.calc(self.local_hostname_fqdn.encode("utf-8")))
```
This line should be
self.csv2_host_id = int(Crc32.calc(db_config['db_host']))

colsond commented on web_frontend/cloudscheduler/csv2/group_views.py:
```diff
@@ -337,6 +339,52 @@ def defaults(request, active_user=None, response_code=0, message=None):
                 return redirect("/group/defaults/")
                 #return render(request, 'csv2/group_defaults.html', {'response_code': 1, 'message': '%s default update/list %s' % (lno(MODID), msg), 'active_user': active_user.username, 'active_group': active_user.active_group, 'user_groups': active_user.user_groups})
 
+			#validate_fqdn
```
This section of code should now simply detect a change in condor host and call the cleanuo routine before pushing the update thru.

SeojLee committed:
Modify where clause for csv2_service_catalog table

SeojLee committed:
Update csv2_host_id calculation to use db_host

SeojLee committed:
Remove FQDN validation and associated queries

Removed unused FQDN validation logic and related database queries.

SeojLee committed:
Optimize deletion of stale entries in database

Refactor deletion logic to use a where clause list for batch deletion.

SeojLee committed:
Delete schema/schema_backup/elephant190.heprc.uvic.ca

SeojLee committed:
Implement cleanup for stale service catalog

Add cleanup_stale_service_catalog call when FQDN changes

SeojLee committed:
Add cleanup_stale_service_catalog function

Implement cleanup function for stale service catalog entries.

SeojLee committed:
Remove cleanup_stale_service_catalog function

Removed the cleanup_stale_service_catalog function and updated imports.

SeojLee committed:
Fix where clause for csv2_service_catalog query

SeojLee committed:
Exclude active IDs from stale entries processing

Filter stale entries to exclude active IDs before processing.

SeojLee committed:
Correct stale entries filtering in view_utils.py

Fix logic to filter stale entries based on active IDs.

SeojLee committed:
Refactor host ID assignment for clarity

SeojLee committed:
Fix indentation for submitted_host_id assignment

SeojLee committed:
Refactor where clause for database query

Commit:
fix indent issues

Commit:
alter stale time

colsond commented on lib/db_config.py:
```diff
@@ -563,26 +564,7 @@ def get_config_by_category(self, categories):
 #-------------------------------------------------------------------------------
 
     def get_host_id_by_fqdn(self, fqdn):
```
I'm not sure if this function is actually called anywhere, you dont use it in the function above and i'm pretty sure it doesn't appear anywhere else in the codebase. Double check but this may not have been in use even before your change.

colsond commented on lib/view_utils.py:
```diff
@@ -83,6 +84,52 @@ def cskv(key_values, opt='dict'):
+            if rc != 0:
+                continue
+            where_list = []
+            for entry in stale_entries:
+                host_id = entry.get(col)
+                if host_id not in active_ids:
+                    where_list.append("%s = %s" % (col, str(host_id)))
```
There should not be any duplicate IDs in the table for condor ids so this is a non issue but its good to keep in mind efficient query building.

SeojLee replied:
Does not apply for the above reason.

colsond commented on lib/view_utils.py:
```diff
@@ -1411,7 +1459,7 @@ def check_value_type(value, fmt):
                         try:
                             current_hostname = socket.gethostbyname(value)
```
current_hostname variable now never used since we encode the url directly instead of the IP.

SeojLee committed:
Remove unused get_host_id_by_fqdn method

Removed unused method get_host_id_by_fqdn from db_config.py.

SeojLee committed:
Remove socket hostname resolution

Remove unused hostname resolution from socket.

SeojLee commented on lib/db_config.py:
```diff
@@ -75,12 +77,11 @@ def __init__(self, db_yaml, categories, db_config_dict=False, db_config_only=Fal
-        self.local_host_id = int(ipaddress.IPv4Address(socket.gethostbyname(socket.gethostname())))
+        self.local_hostname_fqdn = socket.getfqdn()
+        self.local_host_id = int(Crc32.calc(self.local_hostname_fqdn.encode("utf-8")))
+        self.csv2_host_id = int(Crc32.calc(self.local_hostname_fqdn.encode("utf-8")))
```
addressed

SeojLee commented on web_frontend/cloudscheduler/csv2/group_views.py:
```diff
@@ -337,6 +339,52 @@ def defaults(request, active_user=None, response_code=0, message=None):
+
+            rc2, msg2, found_group_list = config.db_query("csv2_groups",where=f"group_name='%s'"%fields.get('group_name'))
+			
+			submitted_fqdn = fields.get('htcondor_fqdn')
```
addressed

SeojLee commented on web_frontend/cloudscheduler/csv2/group_views.py:
```diff
@@ -337,6 +339,52 @@ def defaults(request, active_user=None, response_code=0, message=None):
+                    rc2, msg2, found_group_list = config.db_query("csv2_vms",where=f"group_name='%s'"%fields.get('group_name'))
+                   
+                    current_hostname = found_group_list[0].get('hostname')
+                    if rc== 0:
```
Does not apply as hostname's format is consistent and does not need validation.

SeojLee commented on cloudscheduler/csmain:
```diff
@@ -125,6 +125,46 @@ def available_resources_initialize(config):
+                logging.info(entry)
+                host_id = entry.get(col)
+                if host_id and host_id not in active_ids:
+                    config.db_execute("delete from %s where %s = %s" % (table, col, host_id))
```
addressed

SeojLee commented on cloudscheduler/csmain:
```diff
@@ -125,6 +125,46 @@ def available_resources_initialize(config):
+            cleanup_timeout = 300
+            logging.warning("invalid 'service_cleanup_timeout,' using default: %ss." %cleanup_timeout)
+
+        stale_time = int(time.time()) - cleanup_timeout
```
addressed
